### PR TITLE
Add support for ignore nulls for LEAD, LAG in WindowAggExec

### LINKED ISSUE #9456: Add support for ignore nulls for LEAD, LAG in WindowAggExec.
### Is your feature request related to a problem or challenge?

Currently `LEAD` and `LAG` window functions have `IGNORE NULLS` support when they are calculated from `BoundedWindowAggExec`. However, when they are calculated from `WindowAggExec` they don't have this support.

Currently datafusion gives error: `IGNORE NULLS mode for LAG and LEAD is not supported for WindowAggExec` when `WindowAggExec` is used with ignore nulls option.

As an example test below in the `window.slt` file
```sql
select lag(a, 2, null) ignore nulls over (order by id desc) as x1,
       lag(a, 2, 'def') ignore nulls over (order by id desc) as x2,
       lag(a, 2, null) respect nulls over (order by id desc) as x4,
       lag(a, 2, 'def') respect nulls over (order by id desc) as x5,
       sum(id) over (order by id desc ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as sum_id
from (select 2 id, 'b' a union all select 1 id, null a union all select 3 id, null union all select 4 id, 'x')
```
triggers this error.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

The error is triggered at `evaluate_all` method of the `WindowShiftEvaluator`. This implementation should be updated to support this feature.

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #9456

### CODE REVIEW

File: datafusion/physical-expr/src/window/lead_lag.rs [hunk 1]
```diff
@@ -151,6 +149,64 @@ impl WindowShiftEvaluator {
...
 }
 
+// implement ignore null for evaluate_all
+fn evaluate_all_with_ignore_null(
+    array: &ArrayRef,
+    offset: i64,
+    default_value: &ScalarValue,
+    is_lag: bool,
+) -> Result<ArrayRef, DataFusionError> {
+    let valid_indices: Vec<usize> = (0..array.len())
```
Comment (comphead) [inline datafusion/physical-expr/src/window/lead_lag.rs:159]: there is already a method for that `array.nulls()`

Comment (Lordworms) [reply datafusion/physical-expr/src/window/lead_lag.rs:159]: yes, but in this case I want to get the valid indices?

Comment (comphead) [reply datafusion/physical-expr/src/window/lead_lag.rs:158]: `valid_indices` is everything in array which is not in `array.nulls()`

so `valid_indices.len() == array.len() - array.nulls().len()`

File: datafusion/physical-expr/src/window/lead_lag.rs [hunk 2]
```diff
@@ -151,6 +148,59 @@ impl WindowShiftEvaluator {
...
+        .filter(|&index| array.is_valid(index))
+        .collect();
+
+    let new_array_results: Result<Vec<_>, DataFusionError> = (0..array.len())
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/window/lead_lag.rs:162]: I think following calculations do not change each iteration. Hence, you can move them out to the iteration as below
```rust    
let direction = is_lag ^ (offset > 0);
let offset = if direction {
    offset as usize
} else {
    offset.unsigned_abs() as usize
};

let new_array_results: Result<Vec<_>, DataFusionError> = (0..array.len())
...
...
```

File: datafusion/sqllogictest/test_files/window.slt [hunk 3]
```diff
@@ -4159,15 +4176,34 @@ NULL def NULL def
 NULL def x x
 x x NULL NULL
 
+# LAG window function IGNORE/RESPECT NULLS support with descending order and nondefault offset trigger evaluate_all
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/window.slt:4179]: I think, this test is exactly same version of the test below. If so, it is better to remove one of them.

File: datafusion/physical-expr/src/window/lead_lag.rs (hunk 1 above)
Comment (Lordworms) [reply datafusion/physical-expr/src/window/lead_lag.rs:158]: Thanks

File: datafusion/sqllogictest/test_files/window.slt (hunk 3 above)
Comment (Lordworms) [reply datafusion/sqllogictest/test_files/window.slt:4179]: Sure, I'll remove it

File: datafusion/physical-expr/src/window/lead_lag.rs (hunk 2 above)
Comment (Lordworms) [reply datafusion/physical-expr/src/window/lead_lag.rs:162]: sure

File: datafusion/sqllogictest/test_files/window.slt [hunk 4]
```diff
@@ -4133,18 +4133,35 @@ x x x x x x
...
+# LAG window function IGNORE/RESPECT NULLS support with ascending order and nondefault offset trigger evaluate_all
+query TTTTI
 select lag(a, 2, null) ignore nulls over (order by id) as x1,
        lag(a, 2, 'def') ignore nulls over (order by id) as x2,
        lag(a, 2, null) respect nulls over (order by id) as x4,
-       lag(a, 2, 'def') respect nulls over (order by id) as x5
+       lag(a, 2, 'def') respect nulls over (order by id) as x5,
+       sum(id) over (order by id desc ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as sum_id
 from (select 2 id, 'b' a union all select 1 id, null a union all select 3 id, null union all select 4 id, 'x')
 ----
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/window.slt:4160]: double checked tests result in Spark, evetything is correct 👍

Comment (Lordworms) [reply datafusion/sqllogictest/test_files/window.slt:4160]: Thanks

File: datafusion/sqllogictest/test_files/window.slt [hunk 5]
```diff
@@ -4227,13 +4249,18 @@ NULL def NULL def
...
 # To trigger WindowAggExec, we added a sum window function with all of the ranges.
-statement error Execution error: IGNORE NULLS mode for LAG and LEAD is not supported for WindowAggExec
+query TTTTI
 select lead(a, 2, null) ignore nulls over (order by id desc) as x1,
        lead(a, 2, 'def') ignore nulls over (order by id desc) as x2,
        lead(a, 2, null) respect nulls over (order by id desc) as x4,
        lead(a, 2, 'def') respect nulls over (order by id desc) as x5,
        sum(id) over (order by id desc ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as sum_id
 from (select 2 id, 'b' a union all select 1 id, null a union all select 3 id, null union all select 4 id, 'x')
+----
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/window.slt:4259]: this test is not the same as Spark or Trino.
Their result
```
x1|x2 |x4|x5 |sum_id|
--+---+--+---+------+
  |def|b |b  |    10|
  |def|  |   |    10|
  |def|  |def|    10|
  |def|  |def|    10|
```

Comment (Lordworms) [reply datafusion/sqllogictest/test_files/window.slt:4259]: Sure

**Review by comphead — APPROVED** [review]
thanks @Lordworms lgtm

**Pull request merged**